Orange/widgets/data/owfile.py

```diff
@@ -341,22 +344,34 @@ def apply_domain_edit(self):
         attributes = []
         class_vars = []
         metas = []
-        weight = []
-        places = [attributes, class_vars, metas, weight]
-        cols = [[], [], [], []]  # Xcols, Ycols, Mcols, Wcols
-        for column, (name, tpe, place, vals), orig_var in \
+        places = [attributes, class_vars, metas]
+        cols = [[], [], []]  # Xcols, Ycols, Mcols
+        for column, (name, tpe, place, vals, is_con), (orig_var, orig_plc) in \
             zip(count(), self.editor_model.variables,
-                chain(self.data.domain.variables, self.data.domain.metas)):
+                chain([(att, 0) for att in self.data.domain.attributes],
+                      [(cls, 1) for cls in self.data.domain.class_vars],
+                      [(met, 2) for met in self.data.domain.metas])):
             if place == 3:
                 continue
+            col_data = list(chain(*self.data[:, orig_var]))
+            if orig_plc == 2:
+                col_data = list(chain(*self.data[:, orig_var].metas))
             if name == orig_var.name and tpe == type(orig_var):
                 var = orig_var
+            elif tpe == DiscreteVariable:
+                values = list(str(i) for i in set(col_data))
+                var = tpe(name, values)
+                col_data = [values.index(str(x)) for x in col_data]
             else:
                 var = tpe(name)
             places[place].append(var)
-            cols[place].append(column)  # This assumes that no columns were skipped in the original file!
+            cols[place].append(col_data)
         domain = Domain(attributes, class_vars, metas)
-        # data = Table.from_file(self.loaded_file, domain=domain, columns=cols)
+        data = [d for col in cols for d in col]
+        data_t = [[data[i][j] for i in range(len(data))]
+                  for j in range(len(data[0]))]
```

Comment:
Yes, it is faster with numpy. But from_numpy doesn't allow to build a table with an empty X. Than, maybe I should enable that as well.